src/test/java/com/baidu/hugegraph/unit/concurrent/BarrierEventTest.java

```diff
@@ -112,33 +158,79 @@ public void testSignalAll() throws InterruptedException {
+        CountDownLatch signalLatch = new CountDownLatch(1);
         for (int i = 0; i < waitThreadNum; i++) {
             executorService.submit(() -> {
                 try {
+                    signalLatch.await();
+                    waitLatch.countDown();
                     barrierEvent.await();
-                    result.incrementAndGet();
-                    latch.countDown();
+                    eventCount.incrementAndGet();
                 } catch (InterruptedException e) {
-                    // Do nothing
+                    waitThreadInterruptedCount.incrementAndGet();
                 }
             });
         }
-        Thread signalThread = new Thread(() -> {
+
+        executorService.submit(() -> {
             barrierEvent.signalAll();
-            latch.countDown();
+            signalLatch.countDown();
         });
-        signalThread.start();
-        latch.await();
-        executorService.shutdownNow();
 
-        Assert.assertEquals(waitThreadNum, result.get());
+        executorService.shutdown();
+        executorService.awaitTermination(1L, TimeUnit.SECONDS);
+        Assert.assertEquals(10, eventCount.get());
+        Assert.assertEquals(0, waitThreadInterruptedCount.get());
+    }
+
+    @Test
+    public void testSignalAllByMultiThreadWithSignalLast()
+                throws InterruptedException {
+        BarrierEvent barrierEvent = new BarrierEvent();
+        AtomicInteger eventCount = new AtomicInteger(0);
+        AtomicInteger waitThreadInterruptedCount = new AtomicInteger(0);
+        AtomicInteger signalThreadInterruptedCount = new AtomicInteger(0);
+        int waitThreadNum = 10;
+        ExecutorService executorService =
+                        Executors.newFixedThreadPool(waitThreadNum  + 1);
+        CountDownLatch waitLatch = new CountDownLatch(waitThreadNum);
+        CountDownLatch signalLatch = new CountDownLatch(1);
+        for (int i = 0; i < waitThreadNum; i++) {
+            executorService.submit(() -> {
+                try {
+                    waitLatch.countDown();
+                    barrierEvent.await();
+                    eventCount.incrementAndGet();
+                } catch (InterruptedException e) {
+                    waitThreadInterruptedCount.incrementAndGet();
+                }
+            });
+        }
+
+        executorService.submit(() -> {
+            try {
+                waitLatch.await();
+            } catch (InterruptedException e) {
+                signalThreadInterruptedCount.incrementAndGet();
+            }
+            barrierEvent.signalAll();
+            signalLatch.countDown();
+        });
+        signalLatch.await();
+        executorService.shutdown();
+        executorService.awaitTermination(1L, TimeUnit.SECONDS);
+        Assert.assertEquals(waitThreadNum, eventCount.get());
+        Assert.assertEquals(0, waitThreadInterruptedCount.get());
+        Assert.assertEquals(0, signalThreadInterruptedCount.get());
```

Comment:
add testSignalAllByMultiThreadWithSignalAwaitConcurrent for signalAll and await at the same time